### PR TITLE
Fix CPU cumsum signed zero

### LINKED ISSUE #195851: `torch.cumsum` drops the sign of IEEE `-0.0` on CPU (CUDA and NumPy preserve it)
### 🐛 Describe the bug

`torch.cumsum` on CPU turns IEEE-754 negative zero (`-0.0`) into positive zero. The input has the sign bit set; after `cumsum` it does not. NumPy and PyTorch CUDA keep the sign; `torch.cumprod` on the same CPU tensor also keeps it.

This is a signed-zero semantics issue, not a rounding / tolerance difference under the numerical accuracy notes: `(-0.0) + (-0.0)` is `-0.0` in IEEE-754. `torch.signbit` is the right check here.

Related but different polarity or op (CPU keeps `-0.0`, CUDA drops it, or a different function): #162235, #114569, #181801, #87131. I could not find a `cumsum`-specific report of CPU dropping `-0.0`.

```python
import torch

x = torch.tensor([-0.0, -0.0, 1.0])
print(torch.signbit(x).tolist())                   # [True, True, False]
print(torch.signbit(torch.cumsum(x, 0)).tolist())  # [False, False, False]
```

Expected: CPU `cumsum` preserves `-0.0`, so the second line should also be `[True, True, False]`.

Cross-checks on the same input, torch 2.14.0+cpu:

```python
import numpy as np
np.signbit(np.cumsum(np.array([-0.0, -0.0, 1.0])))  # [True, True, False]
torch.signbit(torch.cumprod(x, 0)).tolist()         # [True, False, False]
```

Same CPU loss on 2.13.0+cu130 and 2.8.0+cu128; CUDA on those builds gives `[True, True, False]`. Also reproduces for `float16` / `bfloat16` / `float32` / `float64`. `sum([-0.0])` is `+0.0` on both devices, so this is specific to `cumsum`.

### Versions

CPU arm: PyTorch 2.14.0+cpu, NumPy 2.5.2, Python 3.12.3, Ubuntu 22.04.5 (x86_64)
CUDA arm: PyTorch 2.13.0+cu130, RTX 5090, driver 580.95.05
Also reproduces on 2.8.0+cu128 (original discovery run).

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01

### PR DESCRIPTION
**Pull request by devnp2007**
Fixes #195851

### Summary

`torch.cumsum` on CPU currently loses the sign of IEEE-754 `-0.0`
because the cumulative sum starts from a `+0.0` initializer.

Initialize the cumsum accumulator from the first input element instead,
then continue accumulating from the second element. This preserves the
signed zero of the first element.

### Testing

- `python -m pytest test/test_reductions.py -k "cumsum"`
- `python -m pytest test/test_reductions.py -k test_cumsum_preserves_signed_zero`

Both pass.

The regression test covers float16, bfloat16, float32, and float64.

### Notes

I investigated this issue after coordinating with the issue author and
@abhinav-phi, who were also looking into the root cause.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01

### CODE REVIEW

**Comment by devnp2007** [comment]
Hi @jgong5 @mingfeima — just following up on this PR after 3 days. The CPU `cumsum` root cause has been traced to the accumulator initialization in `cpu_cum_base_kernel`, and the PR includes the minimal fix plus regression coverage across float16, bfloat16, float32, and float64.

CI is currently green. If you have a chance to review when convenient, I'd really appreciate it. Thanks!

**Review by aditew01 — APPROVED** [review]
Looks good

**Comment by devnp2007** [comment]
Hi @jgong5, just checking in on the review for #196206 when you have a chance.

The PR has all checks passing and has received an approval from @aditew01. The change is limited to the CPU cumsum accumulator initialization and includes regression coverage for float16, bfloat16, float32, and float64.

No rush — I’d appreciate your review when convenient. Thanks!

**Comment by devnp2007** [comment]
Hi @jgong5, just following up once more on #196206.

The PR is still green with all checks passing, and @aditew01 has approved the changes. Since your review is still pending, I’d appreciate it if you could take a look when you have a chance.

Thanks!

**Comment by devnp2007** [comment]
Hi @jgong5, just following up once more on https://github.com/pytorch/pytorch/pull/196206.

The PR is still green with all checks passing, and @aditew01 has approved the changes. Since your review is still pending, I’d appreciate it if you could take a look when you have a chance.

Thanks!
